nexus/db-queries/src/db/column_walker.rs

```diff
@@ -0,0 +1,114 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! CTE utility for iterating over all columns in a table.
+
+use diesel::prelude::*;
+use std::marker::PhantomData;
+
+/// Used to iterate over a tuple of columns ("T").
+///
+/// Diesel exposes "AllColumns" as a tuple, which is difficult to iterate over
+/// -- after all, all the types are distinct. However, each of these types
+/// implements "Column", so we can use a macro to provide a
+/// "convertion-to-iterator" implemenation for our expected tuples.
+pub(crate) struct ColumnWalker<T> {
+    remaining: PhantomData<T>,
+}
+
+impl<T> ColumnWalker<T> {
+    pub fn new() -> Self {
+        Self { remaining: PhantomData }
+    }
+}
+
+macro_rules! impl_column_walker {
+    ( $($column:ident)+ ) => (
+        impl<$($column: Column),+> IntoIterator for ColumnWalker<($($column,)+)> {
+            type Item = &'static str;
+            type IntoIter = std::vec::IntoIter<Self::Item>;
+
+            fn into_iter(self) -> Self::IntoIter {
+                // TODO: don't convert to vec? You'll need to figure
+                // out how to state the type of IntoIter.
+                [$($column::NAME,)+].to_vec().into_iter()
```

Comment:
I would like to turn this into the following

```suggestion
                [$($column::NAME,)+].into_iter()
```

... but this would require changing:

```rust
type IntoIter = std::vec::IntoIter<Self::Item>;
```

Into:

```rust
type IntoIter = std::array::IntoIter<Self::Item, N>;
```

I don't know how to specify the type of "N" here, even though it **should** be knowable -- it's the length of `$column`.

Reply:
I think either of the first two options in https://danielkeep.github.io/tlborm/book/blk-counting.html would be fine. A better approach would be https://github.com/rust-lang/rust/issues/83527 which isn't stable yet.

Reply:
Actually in this case you can just pass in the length as an argument to `impl_column_walker`.

Reply:
Ah, good call, I'll pass the length explicitly.